File: src/pascal/2023/day3/part1.pas
program day3part1;
uses sysutils;

var
file_: text;
prev, curr, next: string;
sum, game: int32;
state: int32;
number: int32;
valid: boolean;

i, b: int32;
c: char;

LINE_LENGTH: int32 = 140;

function is_symbol(str: string; ind: int32): boolean;
begin
    is_symbol := ((not ((ind < 1) or (ind > LINE_LENGTH))) and (('0' > str[ind]) or (str[ind] > '9')) and (str[ind] <> '.'));
end;

function is_ok(prev: string; curr: string; next: string; ind: int32): boolean;
begin
    is_ok := is_symbol(curr, i-1);
    is_ok := (is_ok or is_symbol(curr, i+1));
    is_ok := (is_ok or is_symbol(next, i-1));
    is_ok := (is_ok or is_symbol(next, i));
    is_ok := (is_ok or is_symbol(next, i+1));
    is_ok := (is_ok or is_symbol(prev, i-1));
    is_ok := (is_ok or is_symbol(prev, i));
    is_ok := (is_ok or is_symbol(prev, i+1));
end;


(* Here the main program block starts *)
begin
    { open inp.txt for reading }
    assign(file_, 'resources/2023/day3.txt');
    reset(file_);

    { initialize sum and count to 0 }
    sum := 0;

    { read the first two lines }
    readln(file_, curr);
    readln(file_, next);

    writeln(length(curr));

    writeln(booltostr(true));
    writeln(booltostr(false));

    { writeln(curr); }
    { writeln(next); }

    state := 0;
    number := 0;
    { 0 : waiting for number }
    { 1 : reading number }

    { first loop }
    for i := 1 to LINE_LENGTH do
    begin
        c := curr[i];
        case state of
                {waiting for number}
                0:
                begin
                    {number}
                    if (('0' <= c) and (c <= '9')) then
                    begin
                        {add number to number}
                        val(c, number);
                        valid := false;
                        state := 1;

                        valid := (valid or is_ok(next, curr, next, i-1));

                    end;
                end;
                {reading number and checking true}
                1:
                begin
                    if (('0' <= c) and (c <= '9')) then
                    begin
                        {add number to number}
                        val(c, b);
                        number := number * 10 + b;

                        {look for symbol}

                        valid := (valid or is_ok(next, curr, next, i-1));

                    end
                    else 
                    begin
                        state := 0;

                        if (valid) then
                            sum := sum + number;

                    end;
                        
                end;
        end;
    end;

    { main loop }
    while not eof(file_) do 
    begin

        state := 0;

        prev := curr;
        curr := next;
        readln(file_, next);

        writeln();
        writeln(prev);
        writeln(curr);
        writeln(next);

        for i := 1 to LINE_LENGTH do
        begin
            c := curr[i];
            case state of
                    {waiting for number}
                    0:
                    begin
                        {number}
                        if (('0' <= c) and (c <= '9')) then
                        begin
                            {add number to number}
                            val(c, number);
                            valid := false;
                            state := 1;

                            valid := (valid or is_ok(prev, curr, next, i-1));

                        end;
                    end;
                    {reading number and checking true}
                    1:
                    begin
                        if (('0' <= c) and (c <= '9')) then
                        begin
                            {add number to number}
                            val(c, b);
                            number := number * 10 + b;

                            {look for symbol}

                            valid := (valid or is_ok(prev, curr, next, i-1));

                        end
                        else 
                        begin
                            state := 0;

                            if (valid) then
                                sum := sum + number;

                        end;
                            
                    end;
            end;

            write(state);

        end;

        if (state = 1) then
        begin
            begin

                if (valid) then
                    sum := sum + number;

            end;
        end;

    end;

    state := 0;

    prev := curr;
    curr := next;
    

    for i := 1 to LINE_LENGTH do
    begin
        c := curr[i];
        case state of
                {waiting for number}
                0:
                begin
                    {number}
                    if (('0' <= c) and (c <= '9')) then
                    begin
                        {add number to number}
                        val(c, number);
                        valid := false;
                        state := 1;

                        valid := (valid or is_ok(prev, curr, prev, i-1));

                    end;
                end;
                {reading number and checking true}
                1:
                begin
                    if (('0' <= c) and (c <= '9')) then
                    begin
                        {add number to number}
                        val(c, b);
                        number := number * 10 + b;

                        {look for symbol}

                        valid := (valid or is_ok(prev, curr, prev, i-1));

                    end
                    else 
                    begin
                        state := 0;

                        if (valid) then
                            sum := sum + number;

                    end;
                        
                end;
        end;
    end;

    writeln();
    writeln(sum);
end.

File: src/python/2023/day3/part1.py
print(sum(sum(
    [[m[2] for m in zip(*l) if m[0] and not (m[1] and m[2] == m[3]) and m[2] != -1] for l in
        list(map(lambda x, y: (x, [False] + x, y, [-1] + y),
            [[sum(n) > 0 for n in zip(*m)] for m in zip(
                *list(map(lambda x: [
                        [[False] + q for q in [[False] * len(x[0])] + x], [[False] * len(x[0])] + x, [q[1:] + [False] for q in [[False] * len(x[0])] + x],
                        [[False] + q for q in x], x, [q[1:] + [False] for q in x],
                        [[False] + q for q in x[1:] + [[False] * len(x[0])]],  x[1:] + [[False] * len(x[0])], [q[1:] + [False] for q in x[1:] + [[False] * len(x[0])]]],
                    [[[x not in '0123456789.' for x in l] for l in open("resources/2023/day3.txt").read().split("\n")]] # array of True when a symbol
                ))[0]
            )], # array of True when next to a symbol
            [[max(a) if a[1] != -1 else -1 for a in zip(*b)] for b in zip(
                *list(map(lambda x: [[[-1] + q for q in x], x, [q[1:] + [-1] for q in x]],
                    [[[int("".join(c).strip('.!@#$%^&*()_+-=/')) if c[1] in '0123456789' else -1 for c in zip('.' + l, l, l[1:] + '.')] for l in open("resources/2023/day3.txt").read().split("\n")]]
                ))[0]
            )] # array of -1, or the number.
        ))
    ], []
)))

# sum(sum([[m[2] for m in zip(*l) if m[0] and not (m[1] and m[2] == m[3]) and m[2] != -1] for l in list(map(lambda x, y: (x, [False] + x, y, [-1] + y), [[sum(n) > 0 for n in zip(*m)] for m in zip(*list(map(lambda x: [[[False] + q for q in [[False] * len(x[0])] + x], [[False] * len(x[0])] + x, [q[1:] + [False] for q in [[False] * len(x[0])] + x], [[False] + q for q in x], x, [q[1:] + [False] for q in x], [[False] + q for q in x[1:] + [[False] * len(x[0])]],  x[1:] + [[False] * len(x[0])], [q[1:] + [False] for q in x[1:] + [[False] * len(x[0])]]], [[[x not in '0123456789.' for x in l] for l in open("resources/2023/day3.txt").read().split("\n")]]))[0])], [[max(a) if a[1] != -1 else -1 for a in zip(*b)] for b in zip(*list(map(lambda x: [[[-1] + q for q in x], x, [q[1:] + [-1] for q in x]], [[[int("".join(c).strip('.!@#$%^&*()_+-=/')) if c[1] in '0123456789' else -1 for c in zip('.' + l, l, l[1:] + '.')] for l in open("resources/2023/day3.txt").read().split("\n")]]))[0])]))], []))
